### PR TITLE
inputRange: add a `includeNonApplicable` property

### LINKED ISSUE #539: Add "Not applicable" for the user for range input
# Issue

## Justification
In some of our surveys, we often ask radio input questions that are not applicable to the user. As a result, the user doesn't know how to respond in this situation.

## Description
The user should be able to click on 'not applicable'. We should add a parameter to the widget (notApplicable: boolean). 

## Screenshot
We could do something like this.

image

### PR DESCRIPTION
**Pull request by tahini**
fixes #539

To avoid the participant having to select a position for a question that does not apply to him, there should be a "Non applicable" option that the user can select that will disable the input range. The answer in this case, instead of a `number` will be a `string` with value `na`.

### CODE REVIEW

**Comment by tahini** [comment]
See the screenshot in the issue for what it looks like. 

Do you think the text of the "Non applicable" is ok? should be overridable by the individual surveys? And if so, should each question be able to update it?

Is `includeNonApplicable` a good name for the property?

There may also be some aria to add somewhere to make sure the "Non applicable" is linked to the text of the question somehow.

**Comment by kaligrafy** [comment]
The text should be overridable in the survey config

File: example/demo_survey/src/survey/widgets/end.tsx [hunk 1]
```diff
@@ -343,6 +343,7 @@ export const householdSurveyAppreciation = {
     type: "question",
     path: "household.surveyAppreciation",
     inputType: "slider",
+    includeNonApplicable: true,
     datatype: "number",
```
Comment (samuel-duhaime) [inline example/demo_survey/src/survey/widgets/end.tsx:347]: Mettre en anglais. includeNotApplicable 

Comme dit par @kaligrafy, peux-tu ajouter un notApplicableLabel?

File: packages/evolution-common/src/services/widgets/WidgetConfig.ts [hunk 2]
```diff
@@ -194,6 +194,7 @@ export type InputRangeType = {
     formatLabel?: (value: number, lang: string) => string;
     labels?: I18nData[];
     trackClassName?: string;
+    includeNonApplicable?: boolean;
```
Comment (samuel-duhaime) [inline packages/evolution-common/src/services/widgets/WidgetConfig.ts:197]: Mettre en anglais, includeNotApplicable.

File: packages/evolution-frontend/src/components/inputs/InputCheckbox.tsx [hunk 3]
```diff
@@ -18,7 +18,7 @@ import { CommonInputProps } from './CommonInputProps';
 import { sortByParameters } from './InputChoiceSorting';
 
 export type InputCheckboxProps = CommonInputProps & {
-    value: string;
+    value: string[];
```
Comment (samuel-duhaime) [inline packages/evolution-frontend/src/components/inputs/InputCheckbox.tsx:21]: C'était une erreur d'un autre commit?

File: packages/evolution-frontend/src/components/inputs/InputRange.tsx [hunk 4]
```diff
@@ -15,15 +15,16 @@ import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
...
 export type InputRangeProps = CommonInputProps & {
-    value?: number;
+    value?: number | string;
     inputRef?: React.LegacyRef<ReactInputRange>;
     widgetConfig: InputRangeType;
 };
 
 export const InputRange = (props: InputRangeProps & WithTranslation) => {
-    const [value, setValue] = React.useState<number | Range | undefined>(props.value);
+    const [value, setValue] = React.useState<number | Range | undefined | string>(props.value);
```
Comment (samuel-duhaime) [inline packages/evolution-frontend/src/components/inputs/InputRange.tsx:27]: On veut un string ou mettre 'na' ?

File: packages/evolution-frontend/src/components/inputs/InputRange.tsx [hunk 5]
```diff
@@ -15,15 +15,16 @@ import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
 import { InputRangeType } from 'evolution-common/lib/services/widgets';
 import * as surveyHelper from 'evolution-common/lib/utils/helpers';
 import { CommonInputProps } from './CommonInputProps';
+import InputCheckbox from './InputCheckbox';
 
 export type InputRangeProps = CommonInputProps & {
-    value?: number;
+    value?: number | string;
```
Comment (samuel-duhaime) [inline packages/evolution-frontend/src/components/inputs/InputRange.tsx:21]: On veut un string ou mettre 'na' ?

File: packages/evolution-frontend/src/components/inputs/__tests__/InputRange.test.tsx [hunk 6]
```diff
@@ -143,3 +143,71 @@ describe('Should correctly render InputRange with various parameters', () => {
     });
 
 });
+
+describe('Should correctly render InputRange with non applicable', () => {
```
Comment (samuel-duhaime) [inline packages/evolution-frontend/src/components/inputs/__tests__/InputRange.test.tsx:147]: Mettre en anglais 'not applicable'

**Review by samuel-duhaime — CHANGES_REQUESTED** [review]
Peux-tu rajouter dans le Generator, rajouter un row includeNotApplicable?

File: packages/evolution-frontend/src/components/inputs/InputRange.tsx (hunk 5 above)
Comment (tahini) [reply packages/evolution-frontend/src/components/inputs/InputRange.tsx:21]: Bon point, je vais mettre 'na' seulement

**Comment by tahini** [comment]
> Peux-tu rajouter dans le Generator, rajouter un row includeNotApplicable?

C'est fait

File: packages/evolution-frontend/src/components/inputs/InputCheckbox.tsx (hunk 3 above)
Comment (tahini) [reply packages/evolution-frontend/src/components/inputs/InputCheckbox.tsx:21]: C'est un autre commit entier dans cette PR qui corrige une erreur sur ce type qui existe depuis longtemps...

**Pull request merged**